### ISSUE TITLE
crashes on win10 x64 

### ISSUE THREAD

**Issue by laran002** [pisontechnology/nrf52-desktop-dfu#3]
I am trying this out and have some issues on windows x64. I have compiled the master branch using MSVC19 with some additional debugging output in the NrfDfuServer. The secure DFU on my device is from SDK15, if that has an impact.

This is the output in my terminal:
```
Starting DFU Test!
Initiating scan for 2500 milliseconds...
Data file size: 143
Bin file size: 18788
Start scanning for devices:
Starting Scan!
c48d4d32efcf
  Found: NrfDFU (c48d4d32efcf)
640c29ec2b16
4c875ddd6f8b
  Connected to c48d4d32efcf... initiating streaming...
Running FSM
State update to 1
Running FSM
[WRITE_OPCODE] char-write-req: 0x000f  020000
Received length 3: 0x60 02 01
Calling Notify
Event Received  3
Notified
State update to 2
Running FSM
[WRITE_OPCODE] char-write-req: 0x000f  01018F000000
Received length 3: 0x60 01 01
Calling Notify
Event Received  2
Notified
State update to 3
Running FSM
Calculating checksum of length: 143
128C010A4608011242089BA50610341A02B70120002800300038E49201422408031220E66F5C3FF9CA61596D0DC5A7C04433B937A1F64A7FB91B2C4547CED14D153D55480052040801120010001A40E565DBD14A315D20302BBED187614F5DD97457BF760A7DABA1EE45E26A5823BC6B690602F9BD183D65EAE76956F3C54AA41920282A904C9F3E79384966D08C44
CRC for data to be sent is 0xc168e22b
State update to 4
Running FSM
[WRITE_OPCODE] char-write-req: 0x000f  03
```
And this is where the application crashes....

Please advise

**Comment by laran002** [comment]
After looking more into the code I can see it happens when requesting crc of the uploaded data file. It stops on the request DATAFILE_REQ_CHECKSUM (4). Can there be something with the MTU size, ore any other changes to the secure DFU that may cause this?

**Comment by Sir-Branch** [comment]
Apologies for the delay just got back form traveling! The desktop dfu was tested and developed using SDK 15 so that shouldn't be an issue. 

Can you provide a little bit more context into what chipset you are using and what softdevice version, I have some nrf52 dongles on my end which I can use for testing and reproduction of the error

**Comment by laran002** [comment]
I am using nrf52841 with soft device s132 v. 6.1.1. I have tried to debug it further and it looks like it may actually fail earlier, when the data package is being written to the device on characteristics NORDIC_DFU_PACKET_CHAR.

Is it possible the cause can be the MTU size? I found this line in the code in NrfDfuServerTypes.h
`/TODO: MTU Size will depend on platform (MacOs -.-)/`

**Comment by Sir-Branch** [comment]
The final operation before the code crasher in your environment corresponds to a request for the checksum of the datafile that has been copied over:
```  [WRITE_OPCODE] char-write-req: 0x000f  03 ```

The Device should be responding with the flash page checksum

Not sure what might be causing the checksum to fail, this might be caused by the MTU size as you pointed out. Taken from the [Nordic documentation:](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk5.v15.2.0%2Flib_dfu_transport_ble.html)

> The UUID of the DFU Packet characteristic is 0x0002. This characteristic receives data for Device Firmware Updates as DFU packets.
> 
> The maximum size of each packet is derived from the Att MTU size of the connection. The maximum Att MTU size of the DFU Service is 256 bytes (saved in NRF_SDH_BLE_GATT_MAX_MTU_SIZE), making the maximum size of the DFU Packet characteristic 253 bytes. (3 bytes are used for opcode and handle ID upon writing.) If no larger Att MTU size is negotiated with the DFU Service, the expected maximum size of a write is 20 Bytes. (BLE_GATT_ATT_MTU_DEFAULT minus the 3 bytes used for opcode and handle ID.) For simplicity, the Message sequence charts below show an example where no ATT MTU size was negotiated, so that a write of 20 byte is used.
> 
> Packets must be in little endian (LSB first) order.

I would attempt with MTU 256 and MTU 20 and see if that solves the issue (you'll have to change the define and rebuild the library).

I've attached a diagram of the FSM, incase this provides useful in debugging:
![image](https://user-images.githubusercontent.com/14277350/119278092-180dc480-bbf1-11eb-9c23-4b658ee641c7.png)

**Comment by laran002** [comment]
I have tested with different MTU sizes and I think we are onto something. I created a new pull request https://github.com/pisontechnology/nrf52-desktop-dfu/pull/5 adressing the issue with small MTU_CHUNK sizes.

I have been able to properly use it on both linux (bluez) and windows. However, there is something fishy on windows since eventually an incorrect crc is returned during each and every update.

I have managed to get stable by using the Qt bluetooth implementation ( getting around the nativeble implementation) then the crc is read back correct.

**Comment by Sir-Branch** [comment]
I really appreciate the insight! Happy to hear you were able to make it work.

I agree lowering the MTU Chunk size will make the update more portable to different hardware, and speed really isn't of importance at this moment. 

I'll review the PR tonight and merge it in (on my fork).

Let me know if there's anything else I can assist with!

**Comment by laran002** [comment]
I'll continue my work on porting the NrfDfuServer to Qt and will publish the result when done. The aim here is to make it available on handsets as well.

I have still not tested on OSX since I haven't figured out how tho get the UUID of the device. I tried to scan in different ways without success, any ideas?

**Comment by Sir-Branch** [comment]
What library are you using for scanning on OSX? What results are you seeing when scanning with OSX?

Note the following:

> macOS - MAC Addresses and UUIDs
> In an effort to protect privacy, CoreBluetooth (the underlying macOS Bluetooth API) does not expose the MAC address of a device to a user. Instead, it randomizes the MAC address to a UUID (Universal Unique Identifier) that is exposed to the user. Instead, you will need to scan for devices and find the UUID of the desired device to connect to.

**Comment by laran002** [comment]
Well, I don't have a lib for scanning. I tried the naive approach adding some debug code and inserted a dummy UUID as argument, expecting the application to list available devices for me. Nothing comes out in the terminal on osx :(
```
callback_holder.callback_on_scan_found = [&](NativeBLE::DeviceDescriptor device) {
        std::cout << device.address << std::endl;
        if (is_mac_addr_match(device.address, device_dfu_ble_address)) {
            std::cout << "  Found: " << device.name << " (" << device.address << ")" << std::endl;
            device_found = true;
            device_dfu_ble_address = device.address;
        }
    };
```

**Comment by Sir-Branch** [comment]
Unfortunately I don't have access to a computer with OSX. Were you at least able to carryout DFUs successfully on the Windows and Linux?

**Comment by laran002** [comment]
Status is the following: DFU update works fine on Linux and Windows. On OSX no devices are found during scanning, at least not for me. However, swapping out the nativeble ble back-end  (in my case to Qt) the dfu server code works fine on osx.

I think we can close this issue and possibly open another related to scanning on osx

**Comment by Sir-Branch** [comment]
Awesome thanks for the insight! I won't be able to close the issue due to permissions, put I will pull in your changes to the development branch in my fork and once I fully test (waiting on the nRF52-DK to arrive) will merge into my master branch and do a formal release!
